File: delete-node-in-a-bst.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def deleteNode(self, root, key):
        """
        :type root: TreeNode
        :type key: int
        :rtype: TreeNode
        """
        def findMax(root):
            node = root
            while node.right:
                node = node.right
            return node.val

        def delete(root, key):
            if root is None:
                return None
            elif key < root.val:
                root.left = delete(root.left, key)
                return root
            elif key > root.val:
                root.right = delete(root.right, key)
                return root
            elif root.left is None and root.right is None:
                return None
            elif root.left is None:
                return root.right
            elif root.right is None:
                return root.left
            else:
                maxInLeft = findMax(root.left)
                root.val = maxInLeft
                root.left = delete(root.left, maxInLeft)
                return root

        return delete(root, key)

File: delete-node-in-a-bst.js
function deleteRoot (root) {
  if (root.left && root.right) {
    root.val = root.left.val
    root.left = deleteRoot(root.left)
    return root
  } else {
    return root.left || root.right
  }
}

/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} key
 * @return {TreeNode}
 */
var deleteNode = function(root, key) {
  if (!root) {
    return root
  } else if (key > root.val) {
    root.right = deleteNode(root.right, key)
    return root
  } else if (key < root.val) {
    root.left = deleteNode(root.left, key)
    return root
  } else {
    return deleteRoot(root)
  }
};
